File: packages/typescript/rpc/src/promise.ts
/**
 * RpcPromise - A Promise with pipelining support
 *
 * Enables chaining method calls without waiting for previous results,
 * allowing multiple operations to be batched into a single round trip.
 */

import type { RpcClient } from './client.js';

/**
 * Operation type for pipeline
 */
export interface PipelineOp {
  type: 'call' | 'get';
  method?: string;
  property?: string;
  args?: unknown[];
}

/**
 * A promise that supports pipelining operations
 */
export class RpcPromise<T> extends Promise<T> {
  private _client: RpcClient<unknown>;
  private _ops: PipelineOp[];
  private _capabilityId?: number;

  constructor(
    executor: (resolve: (value: T | PromiseLike<T>) => void, reject: (reason?: unknown) => void) => void,
    client: RpcClient<unknown>,
    ops: PipelineOp[] = [],
    capabilityId?: number
  ) {
    super(executor);
    this._client = client;
    this._ops = ops;
    this._capabilityId = capabilityId;
  }

  /**
   * Get a property from the result
   */
  get<K extends keyof T>(key: K): RpcPromise<T[K]> {
    const newOps = [...this._ops, { type: 'get' as const, property: String(key) }];

    return new RpcPromise<T[K]>(
      (resolve, reject) => {
        this.then(value => {
          if (value && typeof value === 'object') {
            resolve((value as Record<string, unknown>)[String(key)] as T[K]);
          } else {
            reject(new Error(`Cannot get property ${String(key)} of ${typeof value}`));
          }
        }).catch(reject);
      },
      this._client,
      newOps,
      this._capabilityId
    );
  }

  /**
   * Call a method on the result
   */
  call(method: string, ...args: unknown[]): RpcPromise<unknown> {
    const newOps = [...this._ops, { type: 'call' as const, method, args }];

    return new RpcPromise<unknown>(
      (resolve, reject) => {
        this.then(async value => {
          // If value is a capability reference, call through the client
          if (value && typeof value === 'object' && '__capabilityId' in (value as object)) {
            const result = await this._client.callOnCapability(value, method, ...args);
            resolve(result);
          } else if (value && typeof value === 'object') {
            const fn = (value as Record<string, unknown>)[method];
            if (typeof fn === 'function') {
              resolve(fn.apply(value, args));
            } else {
              reject(new Error(`${method} is not a function`));
            }
          } else {
            reject(new Error(`Cannot call method on ${typeof value}`));
          }
        }).catch(reject);
      },
      this._client,
      newOps,
      this._capabilityId
    );
  }

  /**
   * Server-side map operation
   * Maps a function over an array on the server, avoiding N round trips
   */
  map<U>(fn: (item: T extends (infer I)[] ? I : T) => U): RpcPromise<U[]> {
    // Extract the function body for server execution
    const fnStr = fn.toString();

    return new RpcPromise<U[]>(
      (resolve, reject) => {
        this.then(async value => {
          try {
            const result = await this._client.serverMap(value, fnStr, {});
            resolve(result as U[]);
          } catch (error) {
            reject(error);
          }
        }).catch(reject);
      },
      this._client,
      this._ops,
      this._capabilityId
    );
  }

  /**
   * Get the pipeline operations
   */
  getPipelineOps(): PipelineOp[] {
    return [...this._ops];
  }

  /**
   * Get the capability ID if this promise represents a capability
   */
  getCapabilityId(): number | undefined {
    return this._capabilityId;
  }

  /**
   * Create an RpcPromise from a regular promise
   */
  static from<T>(promise: Promise<T>, client: RpcClient<unknown>, ops: PipelineOp[] = [], capId?: number): RpcPromise<T> {
    return new RpcPromise<T>(
      (resolve, reject) => {
        promise.then(resolve).catch(reject);
      },
      client,
      ops,
      capId
    );
  }
}

/**
 * Pipeline builder for batching multiple operations
 */
export class PipelineBuilder {
  private _client: RpcClient<unknown>;
  private _steps: Array<{
    target?: string;
    method: string;
    args: unknown[];
    as?: string;
  }> = [];
  private _lastAs?: string;

  constructor(client: RpcClient<unknown>) {
    this._client = client;
  }

  /**
   * Add a method call to the pipeline
   */
  call(method: string, ...args: unknown[]): this {
    this._steps.push({ method, args });
    return this;
  }

  /**
   * Call a method on a previously named result
   */
  callOn(targetName: string, method: string, ...args: unknown[]): this {
    this._steps.push({ target: targetName, method, args });
    return this;
  }

  /**
   * Name the result of the previous step
   */
  as(name: string): this {
    if (this._steps.length > 0) {
      this._steps[this._steps.length - 1].as = name;
      this._lastAs = name;
    }
    return this;
  }

  /**
   * Execute the pipeline
   */
  async execute(): Promise<Record<string, unknown> & { __last: unknown }> {
    return this._client.executePipeline(this._steps);
  }
}

File: packages/rust/rpc/src/promise.rs
//! RpcPromise - Lazy RPC result with pipelining support.

use crate::client::{ConnectionState, PendingRequest, PipelineStep, RpcMessage};
use crate::error::{RpcError, TransportError};

use serde::de::DeserializeOwned;
use serde_json::Value as JsonValue;
use std::future::Future;
use std::marker::PhantomData;
use std::pin::Pin;
use std::sync::Arc;
use std::task::{Context, Poll};
use tokio::sync::{oneshot, Mutex};
use tokio_tungstenite::tungstenite::Message;

/// A lazy RPC result that supports pipelining.
///
/// `RpcPromise<T>` implements `Future<Output = Result<T, RpcError>>` and
/// allows chaining method calls before awaiting.
///
/// # Example
///
/// ```ignore
/// // Create a promise for a method call
/// let promise: RpcPromise<Counter> = client.promise("makeCounter", vec![json!(42)]);
///
/// // Chain another call (builds pipeline, no network yet)
/// let value_promise: RpcPromise<i32> = promise.call("value", vec![]);
///
/// // Now execute the pipeline
/// let value = value_promise.await?;
/// ```
pub struct RpcPromise<T> {
    /// Connection state for sending requests.
    state: Arc<Mutex<ConnectionState>>,
    /// Request ID (assigned when first step is created).
    id: u64,
    /// Initial method to call.
    method: String,
    /// Arguments for the initial method.
    args: Vec<JsonValue>,
    /// Target stub ID (for calls on capabilities).
    target: Option<String>,
    /// Pipeline steps built up before await.
    pipeline: Vec<PipelineOp>,
    /// Inner state for the future.
    inner: PromiseState,
    /// Type marker.
    _marker: PhantomData<T>,
}

/// A pipeline operation.
#[derive(Clone, Debug)]
struct PipelineOp {
    /// Method to call.
    method: String,
    /// Arguments.
    args: Vec<JsonValue>,
    /// Alias for the result.
    alias: Option<String>,
    /// Target from previous step.
    target_alias: Option<String>,
}

/// Internal state for the promise future.
enum PromiseState {
    /// Not yet started.
    Pending,
    /// Request sent, waiting for response.
    Waiting(oneshot::Receiver<Result<JsonValue, RpcError>>),
    /// Completed or failed.
    Done,
}

impl<T> RpcPromise<T>
where
    T: DeserializeOwned + Send + 'static,
{
    /// Create a new RpcPromise.
    pub fn new(
        state: Arc<Mutex<ConnectionState>>,
        id: u64,
        method: &str,
        args: Vec<JsonValue>,
        target: Option<String>,
    ) -> Self {
        Self {
            state,
            id,
            method: method.to_string(),
            args,
            target,
            pipeline: Vec::new(),
            inner: PromiseState::Pending,
            _marker: PhantomData,
        }
    }

    /// Chain another method call onto this promise.
    ///
    /// This builds a pipeline that will be executed in a single round trip.
    pub fn call<U>(self, method: &str, args: Vec<JsonValue>) -> RpcPromise<U>
    where
        U: DeserializeOwned + Send + 'static,
    {
        let mut pipeline = self.pipeline;
        pipeline.push(PipelineOp {
            method: method.to_string(),
            args,
            alias: None,
            target_alias: None,
        });

        RpcPromise {
            state: self.state,
            id: self.id,
            method: self.method,
            args: self.args,
            target: self.target,
            pipeline,
            inner: PromiseState::Pending,
            _marker: PhantomData,
        }
    }

    /// Chain a method call with an alias.
    pub fn call_as<U>(self, method: &str, args: Vec<JsonValue>, alias: &str) -> RpcPromise<U>
    where
        U: DeserializeOwned + Send + 'static,
    {
        let mut pipeline = self.pipeline;
        pipeline.push(PipelineOp {
            method: method.to_string(),
            args,
            alias: Some(alias.to_string()),
            target_alias: None,
        });

        RpcPromise {
            state: self.state,
            id: self.id,
            method: self.method,
            args: self.args,
            target: self.target,
            pipeline,
            inner: PromiseState::Pending,
            _marker: PhantomData,
        }
    }

    /// Get the pipeline expression as a string (for debugging).
    pub fn expression_string(&self) -> String {
        let mut expr = format!("{}({:?})", self.method, self.args);
        for op in &self.pipeline {
            expr.push_str(&format!(".{}({:?})", op.method, op.args));
        }
        expr
    }

    /// Build the RPC message for this promise.
    fn build_message(&self) -> RpcMessage {
        if self.pipeline.is_empty() {
            // Simple call
            RpcMessage::Call {
                id: self.id,
                method: self.method.clone(),
                args: self.args.clone(),
                target: self.target.clone(),
            }
        } else {
            // Pipeline call
            let mut steps = vec![PipelineStep {
                call: self.method.clone(),
                args: self.args.clone(),
                alias: Some("_0".to_string()),
                target: self.target.clone(),
            }];

            for (i, op) in self.pipeline.iter().enumerate() {
                steps.push(PipelineStep {
                    call: op.method.clone(),
                    args: op.args.clone(),
                    alias: op.alias.clone().or_else(|| Some(format!("_{}", i + 1))),
                    target: op.target_alias.clone().or_else(|| Some(format!("_{}", i))),
                });
            }

            RpcMessage::Pipeline {
                id: self.id,
                steps,
            }
        }
    }
}

impl<T> Future for RpcPromise<T>
where
    T: DeserializeOwned + Send + 'static,
{
    type Output = Result<T, RpcError>;

    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        loop {
            match &mut self.inner {
                PromiseState::Pending => {
                    // Start the request
                    let msg = self.build_message();
                    let id = self.id;
                    let state = self.state.clone();

                    let (tx, rx) = oneshot::channel();
                    self.inner = PromiseState::Waiting(rx);

                    // Spawn a task to send the request
                    tokio::spawn(async move {
                        let result = send_request(state, id, msg).await;
                        let _ = tx.send(result);
                    });

                    // Continue to poll the waiting state
                    continue;
                }
                PromiseState::Waiting(rx) => {
                    match Pin::new(rx).poll(cx) {
                        Poll::Ready(Ok(result)) => {
                            self.inner = PromiseState::Done;
                            match result {
                                Ok(json) => {
                                    match serde_json::from_value(json) {
                                        Ok(value) => return Poll::Ready(Ok(value)),
                                        Err(e) => {
                                            return Poll::Ready(Err(RpcError::Deserialization(
                                                e.to_string(),
                                            )))
                                        }
                                    }
                                }
                                Err(e) => return Poll::Ready(Err(e)),
                            }
                        }
                        Poll::Ready(Err(_)) => {
                            self.inner = PromiseState::Done;
                            return Poll::Ready(Err(RpcError::Canceled));
                        }
                        Poll::Pending => return Poll::Pending,
                    }
                }
                PromiseState::Done => {
                    return Poll::Ready(Err(RpcError::Internal(
                        "Promise already completed".into(),
                    )));
                }
            }
        }
    }
}

/// Send an RPC request and store the pending handler.
async fn send_request(
    state: Arc<Mutex<ConnectionState>>,
    id: u64,
    msg: RpcMessage,
) -> Result<JsonValue, RpcError> {
    let (response_tx, response_rx) = oneshot::channel();

    {
        let mut state = state.lock().await;
        if !state.connected {
            return Err(TransportError::ConnectionClosed.into());
        }

        state.pending.insert(id, PendingRequest { response_tx });

        let json = serde_json::to_string(&msg)?;
        state
            .message_tx
            .send(Message::Text(json))
            .await
            .map_err(|_| TransportError::ConnectionClosed)?;
    }

    // Wait for response
    match response_rx.await {
        Ok(result) => result,
        Err(_) => Err(RpcError::Canceled),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_pipeline_op() {
        let op = PipelineOp {
            method: "test".to_string(),
            args: vec![],
            alias: Some("result".to_string()),
            target_alias: None,
        };
        assert_eq!(op.method, "test");
        assert_eq!(op.alias, Some("result".to_string()));
    }
}
